### PR TITLE
Feature/dds psm api xtypes test

### PR DESCRIPTION
**Pull request by franava**
adding test

### CODE REVIEW

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/702/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/599/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/697/)
* Performance [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Performance%20Tests%20Localhost/166/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/757/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/642/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/753/)
* Performance [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Performance%20Tests%20Localhost/216/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/798/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/673/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/793/)
* Performance [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Performance%20Tests%20Localhost/247/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/799/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/674/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/794/)
* Performance [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Performance%20Tests%20Localhost/248/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/800/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/675/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/795/)
* Performance [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Performance%20Tests%20Localhost/249/)

**Comment by lemunozm** [comment]
Rebase to main xtypes branch

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/817/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/690/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/812/)
* Performance [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Performance%20Tests%20Localhost/264/)

File: test/dds/core/xtypes/myTest.cpp [hunk 1]
```diff
@@ -473,83 +478,85 @@ DynamicData create_dynamic_data(long double pi, StructType& the_struct/*, Struct
...
+    the_data["int32_t"].value<>(int32_t(-5469372));
+    the_data["uint32_t"].value<>(uint32_t(45350234));
+    the_data["int64_t"].value<>(int64_t(-1234523556));
+    the_data["uint64_t"].value<>(uint64_t(1234523556));
+    the_data["float"].value<>(float(3.1415926));
+    the_data["double"].value<>(double(-3.14159264));
+                
+    the_data["long_double"].value<>(pi);
+
+    for(int i = 0; i < 1E1; ++i) // creating "sequence"
```
Comment (LuisGP) [inline test/dds/core/xtypes/myTest.cpp]: What a shitty sequence :P.

File: test/dds/core/xtypes/myTest.cpp [hunk 2]
```diff
@@ -473,83 +478,85 @@ DynamicData create_dynamic_data(long double pi, StructType& the_struct/*, Struct
...
+    {
+        DynamicData tmp_data(inner_struct);
+        tmp_data["inner_string"].string("lay_down_and_cry");
+        tmp_data["inner_float"].value(float(3.1415));
+        for (int j = 0; j < 1E1; ++j) // creating "sequence.inner_sequence_string"
```
Comment (LuisGP) [inline test/dds/core/xtypes/myTest.cpp]: Store these values in a define/var, so it can be changed easily.

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/818/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/691/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/813/)
* Performance [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Performance%20Tests%20Localhost/265/)

File: test/dds/core/xtypes/myTest.cpp [hunk 3]
```diff
@@ -558,46 +565,780 @@ TEST (DynamicData, cascade_construction)
...
+{
+    {
+        DynamicData dd1(primitive_type<uint8_t>());
+        DynamicData dd2(primitive_type<uint8_t>());
+        dd1.value(uint8_t(15));
+        dd2.value(uint8_t(15));
+        EXPECT_EQ(dd1, dd2);
+        dd2.value(uint8_t(16));
+        EXPECT_NE(dd1, dd2);
+    }
```
Comment (LuisGP) [inline test/dds/core/xtypes/myTest.cpp]: Should something like the following pass the test? I think so.
```cpp
    {
        DynamicData dd1(primitive_type<uint32_t>());
        DynamicData dd2(primitive_type<int32_t>());
        dd1.value(uint32_t(15));
        dd2.value(int32_t(15));
        EXPECT_EQ(dd1, dd2);
        dd2.value(int32_t(16));
        EXPECT_NE(dd1, dd2);
    }
```
As the stored values are the same. The same logic should apply with each compatible primitive types.

File: test/dds/core/xtypes/myTest.cpp (hunk 1 above)
Comment (franava) [reply test/dds/core/xtypes/myTest.cpp]: 1E1 is 10, not 1

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/819/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/692/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/814/)
* Performance [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Performance%20Tests%20Localhost/266/)

File: test/dds/core/xtypes/myTest.cpp (hunk 2 above)
Comment (franava) [reply test/dds/core/xtypes/myTest.cpp]: I'm taking the assumption that this should not be a case.
if u wanna change the struct of a single structure, is much more useful changing the for() line than changing some define value at the beginning of the file

File: test/dds/core/xtypes/myTest.cpp [hunk 4]
```diff
@@ -317,151 +318,155 @@ TEST (DynamicData, primitive_types)
...
+        d["sequence_1"].push('a');
+        d["sequence_1"].push('b');
+        d["sequence_1"].push('c');
+        d["sequence_1"].push('d');
+        d["sequence_1"].push('e');
+        d["sequence_2"].push(d["sequence_1"]);
+    }                
 }
 
 TEST (DynamicData, test_just_for_luis)
```
[the remark this answers is not in the document]
Comment (franava) [reply test/dds/core/xtypes/myTest.cpp]: xD

File: test/dds/core/xtypes/myTest.cpp [hunk 5]
```diff
@@ -558,46 +565,780 @@ TEST (DynamicData, cascade_construction)
...
+        d1["array"][i].value<uint32_t>(42);
+        d1["sequence"].push<uint32_t>(42);
+        d1["seqstring"].push<string>("ci sono anche io");
+    }
+    vector<char> v = d1["string"].as_vector<char>();
+    vector<string> vv = d1["seqstring"].as_vector<string>();
+    vector<uint32_t> t = d1["array"].as_vector<uint32_t>();
+    vector<uint32_t> s = d1["sequence"].as_vector<uint32_t>();
+    
+    cout << v << endl;
```
[the remark this answers is not in the document]
Comment (franava) [reply test/dds/core/xtypes/myTest.cpp]: this was only there for me to check while i was writing the tests.

File: test/dds/core/xtypes/myTest.cpp [hunk 6]
```diff
@@ -558,46 +565,780 @@ TEST (DynamicData, cascade_construction)
...
+    cout << endl ;
+    DynamicData dd(d, s2) ;
+
+    cout << "secondo: " << endl ;
+    for(int i = 0 ; i < 10 ; ++i)
+    {
+        cout << '(' << dd[i].value<uint16_t>() << ')' ; 
+    }
+    cout << endl;
+    EXPECT_EQ(10, dd.size()) ;
```
[the remark this answers is not in the document]
Comment (franava) [reply test/dds/core/xtypes/myTest.cpp]: done

File: test/dds/core/xtypes/myTest.cpp [hunk 7]
```diff
@@ -558,46 +565,780 @@ TEST (DynamicData, cascade_construction)
...
+    StringType t(10);
+    
+    EXPECT_EQ(s.is_compatible(t) , t.is_compatible(s)) ;
+    EXPECT_EQ(TypeConsistency::IGNORE_STRING_BOUNDS, t.is_compatible(s)) ;
+    
+    DynamicData d(s);
+    d.string("12345678901234567890");
+    
+    DynamicData e(d,t) ;
+    EXPECT_EQ(10, e.size()); //still pending investigation
```
[the remark this answers is not in the document]
Comment (franava) [reply test/dds/core/xtypes/myTest.cpp]: done

File: test/dds/core/xtypes/myTest.cpp [hunk 8]
```diff
@@ -558,46 +565,780 @@ TEST (DynamicData, cascade_construction)
...
+    StringType t(20);
+    
+    EXPECT_EQ(s.is_compatible(t) , t.is_compatible(s)) ;
+    EXPECT_EQ(TypeConsistency::IGNORE_STRING_BOUNDS, t.is_compatible(s)) ;
+    
+    DynamicData d(s);
+    d.string("1234567890");
+    
+    DynamicData e(d,t) ;
+    EXPECT_EQ(10, e.size());
```
[the remark this answers is not in the document]
Comment (franava) [reply test/dds/core/xtypes/myTest.cpp]: done

File: test/dds/core/xtypes/myTest.cpp [hunk 9]
```diff
@@ -558,46 +565,780 @@ TEST (DynamicData, cascade_construction)
...
+    WStringType t(10);
+    
+    EXPECT_EQ(s.is_compatible(t) , t.is_compatible(s)) ;
+    EXPECT_EQ(TypeConsistency::IGNORE_STRING_BOUNDS, t.is_compatible(s)) ;
+    
+    DynamicData d(s);
+    d.wstring(L"12345678901234567890");
+    
+    DynamicData e(d,t) ;
+    EXPECT_EQ(10, e.size()); //still pending investigation
```
[the remark this answers is not in the document]
Comment (franava) [reply test/dds/core/xtypes/myTest.cpp]: done

File: test/dds/core/xtypes/myTest.cpp [hunk 10]
```diff
@@ -558,46 +565,780 @@ TEST (DynamicData, cascade_construction)
...
+    WStringType t(20);
+    
+    EXPECT_EQ(s.is_compatible(t) , t.is_compatible(s)) ;
+    EXPECT_EQ(TypeConsistency::IGNORE_STRING_BOUNDS, t.is_compatible(s)) ;
+    
+    DynamicData d(s);
+    d.wstring(L"1234567890");
+    
+    DynamicData e(d,t) ;
+    EXPECT_EQ(10, e.size());
```
[the remark this answers is not in the document]
Comment (franava) [reply test/dds/core/xtypes/myTest.cpp]: done

File: test/dds/core/xtypes/myTest.cpp [hunk 11]
```diff
@@ -558,46 +565,780 @@ TEST (DynamicData, cascade_construction)
...
+    cout << endl;
+    DynamicData dd(d, s2) ;
+
+    cout << "secondo: " << endl ;
+    for(size_t i = 0 ; i < dd.size() ; ++i)
+    {
+        cout << '(' << dd[i].value<uint16_t>() << ')';
+    }
+    cout << endl; 
+    EXPECT_EQ(10, dd.size()) ;
```
[the remark this answers is not in the document]
Comment (franava) [reply test/dds/core/xtypes/myTest.cpp]: done

File: test/dds/core/xtypes/myTest.cpp [hunk 12]
```diff
@@ -289,25 +291,24 @@ TEST (DynamicData, primitive_types)
...
-    d["char32_t"].value(4294967290);
+    DynamicData d(st);
+    
+    d["bool"].value<bool>(true);
+    d["uint8_t"].value<uint8_t>(250);
+    d["int16_t"].value<int16_t>(-32760);
+    d["uint16_t"].value<uint16_t>(65530); 
+    d["int32_t"].value<int32_t>(-2147483640); 
+    d["uint32_t"].value<uint32_t>(4294967290);
+    d["int64_t"].value<int64_t>(-9223372036854775800);
```
[the remark this answers is not in the document]
Comment (franava) [reply test/dds/core/xtypes/myTest.cpp]: done

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/823/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/694/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/818/)
* Performance [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Performance%20Tests%20Localhost/268/)

File: test/dds/core/xtypes/myTest.cpp [hunk 13]
```diff
@@ -473,83 +478,85 @@ DynamicData create_dynamic_data(long double pi, StructType& the_struct/*, Struct
...
-								Member("sequence", SequenceType(inner_struct)));
-				
-				DynamicData the_data(the_struct) ;
-				the_data["bool"].value(true) ;
-    the_data["uint8_t"].value(uint8_t(230)) ;
-				the_data["int16_t"].value<>(int16_t(-784));
-				the_data["uint16_t"].value<>(uint16_t(784));
-				the_data["int32_t"].value<>(int32_t(-5469372));
-				the_data["uint32_t"].value<>(uint32_t(45350234)) ;
-				the_data["int64_t"].value<>(int64_t(-1234523556));
```
[the remark this answers is not in the document]
Comment (franava) [reply test/dds/core/xtypes/myTest.cpp]: done

File: test/dds/core/xtypes/myTest.cpp [hunk 14]
```diff
@@ -473,83 +478,85 @@ DynamicData create_dynamic_data(long double pi, StructType& the_struct/*, Struct
...
-
+    long double pi = 3.14159265358979323846;
+    StructType the_struct("the_struct");
+    StructType inner_struct("inner_struct");
+    StructType second_inner_struct("second_inner_struct");
+    
+    DynamicData the_data = create_dynamic_data(pi, the_struct, inner_struct, second_inner_struct);
 
     debug(false);
     EXPECT_EQ(45350234, the_data["uint32_t"].value<uint32_t>());
```
[the remark this answers is not in the document]
Comment (franava) [reply test/dds/core/xtypes/myTest.cpp]: done

File: test/dds/core/xtypes/myTest.cpp [hunk 15]
```diff
@@ -558,46 +565,780 @@ TEST (DynamicData, cascade_construction)
...
+
+    EXPECT_EQ(TypeConsistency::IGNORE_ARRAY_BOUNDS , the_str.is_compatible(other_str));
+    EXPECT_EQ(TypeConsistency::IGNORE_TYPE_SIGN, the_str.is_compatible(another_str));
+    EXPECT_EQ(TypeConsistency::IGNORE_ARRAY_BOUNDS, other_str.is_compatible(the_str));
+    EXPECT_EQ(TypeConsistency::IGNORE_ARRAY_BOUNDS | TypeConsistency::IGNORE_TYPE_SIGN, other_str.is_compatible(another_str));
+    EXPECT_EQ(TypeConsistency::IGNORE_TYPE_SIGN, another_str.is_compatible(the_str));
+    EXPECT_EQ(TypeConsistency::IGNORE_ARRAY_BOUNDS | TypeConsistency::IGNORE_TYPE_SIGN, another_str.is_compatible(other_str));
+}
+
+TEST (DynamicData, testing_equality_check_primitive_type)
```
[the remark this answers is not in the document]
Comment (franava) [reply test/dds/core/xtypes/myTest.cpp]: done

File: test/dds/core/xtypes/myTest.cpp [hunk 16]
```diff
@@ -558,46 +565,780 @@ TEST (DynamicData, cascade_construction)
...
+        cout << '(' << d[i].value<uint16_t>() << ')';
+    }
+    cout << '\n';
+    DynamicData e(d,b);
+    for(size_t i = 0 ; i < e.size() ; ++i)
+    {
+        cout << '(' << e[i].value<uint16_t>() << ')'; 
+    }
+    cout << '\n';
+    EXPECT_EQ(TypeConsistency::IGNORE_ARRAY_BOUNDS, b.is_compatible(a)) ;
```
[the remark this answers is not in the document]
Comment (franava) [reply test/dds/core/xtypes/myTest.cpp]: done

File: test/dds/core/xtypes/myTest.cpp [hunk 17]
```diff
@@ -558,46 +565,780 @@ TEST (DynamicData, cascade_construction)
...
```
[the remark this answers is not in the document]
Comment (franava) [reply test/dds/core/xtypes/myTest.cpp]: done

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/825/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/696/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/820/)
* Performance [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Performance%20Tests%20Localhost/270/)

File: test/dds/core/xtypes/myTest.cpp [hunk 18]
```diff
@@ -558,46 +565,780 @@ TEST (DynamicData, cascade_construction)
...
+            Member("seq", seq)).add_member(
+            Member("arr", arr)).add_member(
+            Member("primitive", primitive_type<wchar_t>()));
+    
+    b.add_member(
+            Member("string", string)).add_member(
+            Member("seq", seq)).add_member(
+            Member("arr", arr));
+
+    EXPECT_EQ(TypeConsistency::IGNORE_MEMBERS , a.is_compatible(b));
```
[the remark this answers is not in the document]
Comment (franava) [reply test/dds/core/xtypes/myTest.cpp]: done

File: test/dds/core/xtypes/myTest.cpp [hunk 19]
```diff
@@ -289,179 +247,176 @@ TEST (DynamicData, primitive_types)
...
+        d["sequence_1"].push('a');
+        d["sequence_1"].push('b');
+        d["sequence_1"].push('c');
+        d["sequence_1"].push('d');
+        d["sequence_1"].push('e');
+        d["sequence_2"].push(d["sequence_1"]);
+    }                
 }
 
 TEST (DynamicData, test_just_for_luis)
```
[the remark this answers is not in the document]
Comment (franava) [reply test/dds/core/xtypes/myTest.cpp]: removed

**Pull request merged**